In [1]:
import numpy as np 
import pandas as pd

In [2]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv', 'USA.csv', 'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                'USA1.csv', 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv', 
                                 'Italy.csv', 'South_Korea.csv', 'Nigeria.csv',
                                'Colombia.csv', 'Paraguay.csv', 'Russia.csv',
                                'Algeria.csv', 'Chile.csv', 'Yugoslavia.csv',
                                'Scotland.csv', 'Romania.csv', 'Greece.csv',
                                'Honduras.csv', 'Austria.csv', 'Bulgaria.csv', 
                                 'IvoryCoast.csv', 'SouthAfrica.csv', 'RepIreland.csv']))



In [3]:
# 158 with wc 2022 teams

In [4]:
df.shape

(238, 17)

In [5]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [6]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.622869   0.078594      0.347687  0.186192   
xg_z             0.622869  1.000000   0.376119      0.659696  0.321120   
crosses_z        0.078594  0.376119   1.000000      0.642130  0.256452   
boxtouches_z     0.347687  0.659696   0.642130      1.000000  0.372202   
passes_z         0.186192  0.321120   0.256452      0.372202  1.000000   
progpasses_z     0.050513  0.121984   0.276849      0.245929  0.751306   
takeons_z        0.266519  0.357251   0.139513      0.236681  0.383811   
progruns_z       0.187630  0.377608   0.259775      0.380923  0.575545   
tackles_z        0.216870  0.362212   0.244300      0.392027  0.737638   
interceptions_z  0.169602  0.278534   0.231632      0.285272  0.743981   
clearances_z    -0.178897 -0.380202  -0.251982     -0.419246 -0.113066   
blocks_z         0.022937  0.030171   0.010695      0.097848  0.382781   
aerials_z        0.045770  0.033487   0.233689      0.231686  0.000264   
fouls_z         -0.094834 -0.036656   0.064459      0.025010  0.071775   
fouled_z         0.057350  0.170568   0.193971      0.181071  0.295560   
nsxg_z           0.373312  0.641896   0.673878      0.860980  0.484094   
results         -0.597229 -0.292449   0.028692     -0.187746 -0.133468   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.050513   0.266519    0.187630   0.216870   
xg_z                 0.121984   0.357251    0.377608   0.362212   
crosses_z            0.276849   0.139513    0.259775   0.244300   
boxtouches_z         0.245929   0.236681    0.380923   0.392027   
passes_z             0.751306   0.383811    0.575545   0.737638   
progpasses_z         1.000000   0.164441    0.307472   0.515722   
takeons_z            0.164441   1.000000    0.551726   0.376918   
progruns_z           0.307472   0.551726    1.000000   0.511843   
tackles_z            0.515722   0.376918    0.511843   1.000000   
interceptions_z      0.561366   0.374410    0.403296   0.694638   
clearances_z        -0.030842  -0.085508   -0.069473  -0.197803   
blocks_z             0.352512   0.158833    0.264128   0.260157   
aerials_z            0.030705  -0.104707   -0.069694   0.053300   
fouls_z              0.215038   0.127382    0.039159   0.025422   
fouled_z             0.173777   0.338699    0.254810   0.296792   
nsxg_z               0.359863   0.394653    0.471431   0.484239   
results             -0.029293  -0.164389   -0.190845  -0.173461   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.169602     -0.178897  0.022937   0.045770 -0.094834   
xg_z                    0.278534     -0.380202  0.030171   0.033487 -0.036656   
crosses_z               0.231632     -0.251982  0.010695   0.233689  0.064459   
boxtouches_z            0.285272     -0.419246  0.097848   0.231686  0.025010   
passes_z                0.743981     -0.113066  0.382781   0.000264  0.071775   
progpasses_z            0.561366     -0.030842  0.352512   0.030705  0.215038   
takeons_z               0.374410     -0.085508  0.158833  -0.104707  0.127382   
progruns_z              0.403296     -0.069473  0.264128  -0.069694  0.039159   
tackles_z               0.694638     -0.197803  0.260157   0.053300  0.025422   
interceptions_z         1.000000     -0.110541  0.305133   0.025863  0.142297   
clearances_z           -0.110541      1.000000  0.238758   0.266221  0.210544   
blocks_z                0.305133      0.238758  1.000000   0.038173  0.176755   
aerials_z               0.025863      0.266221  0.038173   1.000000  0.109819   
fouls_z                 0.142297      0.210544  0.176755   0.109819  1.000000   
fouled_z                0.309954      0.036525  0.067899   0.024862  0.258934   
nsxg_z                  0.412275     -0.414737  0.134388   0.150250  0.068667   
results                -0.103971      0.078878 -0.057528  -0.037119  0.116708   

                 fo

In [7]:
df.shape

(238, 17)

In [8]:
df.duplicated().value_counts()

False    238
dtype: int64

In [9]:
df.duplicated().value_counts()

False    238
dtype: int64

# convert to csv

In [10]:
df.to_csv('WC-stats.csv', index=False)


In [11]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [12]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [13]:
y = df['results']

In [14]:
X = df.drop(columns=['results'])

In [15]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [16]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.40108458443077766

In [17]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [18]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 1.003 +/- 0.086
    xg_z    0.130 +/- 0.029
    progruns_z0.054 +/- 0.014
    fouled_z0.032 +/- 0.012
    fouls_z 0.018 +/- 0.008
neg_mean_absolute_percentage_error
    goals_z 0.689 +/- 0.077
    fouls_z 0.033 +/- 0.012
    fouled_z0.033 +/- 0.014
    tackles_z0.020 +/- 0.009
neg_mean_squared_error
    goals_z 75.653 +/- 6.467
    xg_z    9.836 +/- 2.194
    progruns_z4.072 +/- 1.072
    fouled_z2.392 +/- 0.893
    fouls_z 1.385 +/- 0.631


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.